### PR TITLE
feat: implement namespace APIs

### LINKED ISSUE #67: Support namespace API
It's convenient to put/get keys in a "namespace" for separating different usage, or it's even required to smoothly write concurrent tests share the same etcd cluster.

This concept, in ZooKeeper, is named as Chroot. I don't find it in etcd but it should be a client-side feature. I create this issue for discussing whether it's desired to implement in this crate.

The API should affect only the constructors for accepting a chroot parameter/options:

```rust
Client::connect(endpoints, chroot /* String */, options);
Client::connect(endpoints, options); // chroot in options
```

... and all the keys pass to this client will be prepended with the "chroot".

### PR DESCRIPTION
**Pull request by tisonkun**
This closes #67.

Still WIP - for preview on the impl direction

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -160,3 +161,16 @@ pub mod proto {
         LockResponse as PbLockResponse, UnlockResponse as PbUnlockResponse,
     };
 }
+
+fn take_mut<T, F>(mut_ref: &mut T, closure: F)
```
Comment (tisonkun) [inline src/lib.rs:165]: This is from https://github.com/Sgeo/take_mut to largely simplify code.

File: src/namespace/kv.rs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        Ok(resp)
+    }
+
+    pub async fn delete(
+        &mut self,
+        key: impl Into<Vec<u8>>,
+        options: Option<DeleteOptions>,
+    ) -> Result<DeleteResponse> {
+        let (key, _) = prefix_internal(&self.pfx, key.into(), vec![]);
+        // TODO (@tisonkun) prefix range_end in options
```
Comment (tisonkun) [inline src/namespace/kv.rs:54]: DeleteOptions make take_mut but the abstraction is strange a bit.

Others are implemented.

**Comment by davidli2010** [comment]
I will take a look at this over the weekend.

**Comment by tisonkun** [comment]
@davidli2010 thank you and as a kindly remind :D

File: src/namespace/mod.rs [hunk 3]
```diff
@@ -0,0 +1,66 @@
+mod kv;
+mod lease;
+
+pub use kv::KvClientPrefix;
+pub use lease::LeaseClientPrefix;
+
+fn prefix_with(pfx: &[u8], key: &mut Vec<u8>) {
+    // HACK - this is effectively: key = pfx.append(key)
+    let _ = key.splice(..0, pfx.to_vec()).collect::<Vec<_>>();
```
Comment (davidli2010) [inline src/namespace/mod.rs:9]: this line has 2 unnecessary memory allocation

File: src/namespace/kv.rs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        resp
+    }
+
+    fn strip_prefix_key(&self, mut key: Vec<u8>) -> Vec<u8> {
+        debug_assert!(
+            key.starts_with(&self.pfx),
+            "{key:?} does not start with {:?}",
+            self.pfx
+        );
+        key.split_off(self.pfx.len())
```
Comment (davidli2010) [inline src/namespace/kv.rs:165]: `split_off` has memory allocation that is not necessary for stripping prefix key.

**Review by davidli2010 — CHANGES_REQUESTED** [review]
I took a quick look at the PR and the namespace APIs were not complicated, but there was too much unnecessary memory allocation. I think we can implement the method of adding and stripping prefix on Vec, while avoiding multiple allocations of memory, and then other code can implement from there:

```rust
trait VecExt {
    fn push_prefix(&mut self, prefix: &[u8]);
    fn strip_prefix(&mut self, prefix: &[u8]);
}

impl VecExt for Vec<u8> {
    fn push_prefix(&mut self, prefix: &[u8]) {
        if prefix.is_empty() {
            return;
        }

        let prefix_len = prefix.len();
        self.reserve(prefix_len);
        unsafe {
            let self_ptr = self.as_mut_ptr();
            std::ptr::copy(self_ptr, self_ptr.add(prefix_len), self.len());
            std::ptr::copy_nonoverlapping(prefix.as_ptr(), self_ptr, prefix_len);
            self.set_len(self.len() + prefix_len);
        }
    }

    fn strip_prefix(&mut self, prefix: &[u8]) {
        if prefix.is_empty() {
            return;
        }

        if self.starts_with(prefix) {
            let prefix_len = prefix.len();
            let new_len = self.len() - prefix_len;
            unsafe {
                let self_ptr = self.as_mut_ptr();
                std::ptr::copy(self_ptr.add(prefix_len), self_ptr, new_len);
                self.set_len(new_len);
            }
        }
    }
}
```

Then we can implement stripping prefix like this:

```rust
impl PutResponse {
    fn strip_prev_kv_prefix(&mut self, prefix: &[u8]) {
        self.prev_kv.as_mut().map(|kv| { kv.key.strip_prefix(prefix);  });
    }
}
```

**Comment by tisonkun** [comment]
@davidli2010 Thanks for your advice! I add the helpers and use them.

Although, I don't extend `strip_prev_kv_prefix` to existing requests but keep using the `take_mut` flavor.

**Comment by davidli2010** [comment]
`take_mut` takes the key and write back after modification. IMO, it's unnecessary.

**Comment by tisonkun** [comment]
```
impl PutResponse {
    fn strip_prev_kv_prefix(&mut self, prefix: &[u8]) {
        self.prev_kv.as_mut().map(|kv| { kv.key.strip_prefix(prefix);  });
    }
}
```

This way is too intrusive and we may have a strange dependency graph for `strip_prefix_xxx_response`.

> take_mut takes the key and write back after modification. IMO, it's unnecessary.

You may try to refactor the code as you wish and see. I don't have a better idea to do now 📦 

And IIUC the current `take_mut` way doesn't introduce heavy operation.

**Comment by davidli2010** [comment]
I refactored the code, you can see it in https://github.com/etcdv3/etcd-client/tree/namespace.

**Comment by tisonkun** [comment]
@davidli2010 your modification looks good, although I may treat them as personal preference differences.

Thanks for your review and refactor! Please go ahead directly updating this PR or use your branch instead :D

**Comment by davidli2010** [comment]
I refactored the code, you can see it in https://github.com/etcdv3/etcd-client/tree/namespace.

> @davidli2010 your modification looks good, although I may treat them as personal preference differences.
> 
> Thanks for your review and refactor! Please go ahead directly updating this PR or use your branch instead :D

I pushed the refactored code to this PR. Can you add some basic test cases and example for the namespace APIs?

**Comment by davidli2010** [comment]
> @davidli2010 your modification looks good, although I may treat them as personal preference differences.
> 
> Thanks for your review and refactor! Please go ahead directly updating this PR or use your branch instead :D

I pushed the refactored code to this PR. Can you add some basic test cases and example for the namespace APIs?

**Comment by tisonkun** [comment]
Thank you! Upstream doesn't have test coverage IIRC. But I can prepare some.

I wonder if we have an integration tests framework already, or shall I write one.

**Comment by davidli2010** [comment]
You can see test cases of client in https://github.com/etcdv3/etcd-client/blob/f5d427b0e0653fac0a8a3dd5d3d6fafd59a3d366/src/client.rs#L832

I think you can add test cases of namespace APIs in `src/namespace/mod.rs` or `tests/namespace.rs`.

**Comment by davidli2010** [comment]
I have splitted test cases to `tests` directory, please see https://github.com/etcdv3/etcd-client/pull/72. You can add namespace test cases in `tests/namespace.rs`, use `testing::get_client` to get client.

**Comment by tisonkun** [comment]
@davidli2010 Updated.

The upstream contains a test for putget and for watch. We may support watch later.

**Review by davidli2010 — CHANGES_REQUESTED** [review]

**Review by davidli2010 — APPROVED** [review]

**Pull request merged**